flink-core/src/main/java/org/apache/flink/api/common/distributions/CommonRangeBoundaries.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.common.distributions;
+
+import org.apache.flink.api.common.typeutils.TypeComparator;
+
+public class CommonRangeBoundaries<T> implements RangeBoundaries<T> {
+	private Object[][] boundaries;
+
+	public CommonRangeBoundaries(Object[][] boundaries) {
+		this.boundaries = boundaries;
+	}
+
+	@Override
+	public int getRangeIndex(T record, TypeComparator<T> typeComparator) {
+		return binarySearch(record, typeComparator);
+	}
+
+	// Search the range index of input record.
+	private int binarySearch(T record, TypeComparator<T> typeComparator) {
+		int low = 0;
+		int high = this.boundaries.length - 1;
+		TypeComparator[] flatComparators = typeComparator.getFlatComparators();
```

Comment:
Can you add the `TypeComparator` to the constructor (it is tied to the boundaries anyways).
Then we can initialize `flatComparators` and `keys` once and reuse them in all `binarySearch()` calls.